documentation/design-docs/ipc-protocol.md

```diff
@@ -0,0 +1,208 @@
+
+```c++
+enum class DiagnosticMessageType : uint32_t
+{
+    // EventPipe
+    StartEventPipeTracing = 1024, // To file
+    StopEventPipeTracing,
+    CollectEventPipeTracing, // To IPC
+};
+
+struct MessageHeader
+{
+    DiagnosticMessageType RequestType;
+    uint32_t Pid;
+};
+```
+
+```
+runtime <- client : MessageHeader { CollectEventPipeTracing }
+    error? -> no response
+runtime -> client : session ID 
+runtime -> client : event stream
+
+...
+
+runtime <- client : stop command
+runtime -> client : session id and stops
+```
+-----
+
+## Proposed Implementation (NEW)
+
+The IPC protocol will become a multi-purpose protocol to be used for communication to and from command servers running in the dotnet core runtime.  EventPipe is a good example of a service that will operate over this protocol.
+
+Connection (general):
+
+![](ipc-protocOl-genericflow.png)
+
+```
+runtime <- client : [ header ][ optional payload ]
+runtime -> client : [ header ][ optional payload ]
+optional
+    runtime <- client : [ optional stream of command specific data ]
+    ------or------
+    runtime -> client : [ optional stream of command specific data ]
+```
+
+EventPipe flow:
+```
+runtime <- client : [ ver; size; start_stream_command ][ stream config struct ]
+runtime -> client : [ ver; size; stream_started_response ][ stream started struct ]
+runtime -> client : [ stream of netperf data ]
+```
+
+### Transport considerations
+
+* All data will be sent little-endian
+* All commands will be sent in a packet with the defined header
+* Command specific data can be sent after a command ACK
+
+
+### Errors
+* runtime: 
+  * read failure -> end connection; no response
+  * unknown protocol -> ERROR response // TODO: flesh out errors
+* Client:
+  * no response -> simply have a timeout
+  * unkown protocol error -> fix your client
+
+### Headers
+
+#### Requirements:
+* small - shouldn't affect write times or read times
+* simple - flat and blittable
+* descriptive - self describing
+* extensible - ability for us to add functionality without necessarily incrementing protocol version
+
+Every packet will start with a header and every header will start with a version magic number and a size.  This makes it easy to read the data directly out of the stream no matter the size or contents of the packet.
+
+> Taking inspiration from Mono's Soft-Debugger protocol, I am splitting command codes into command sets and commands, allowing command codes to be reused between command sets, e.g., 0x01,0x01 and 0x02,0x01 would be different commands despite sharing the same explicit value for the command code.
+
+```c
+// size = 16 + 2 + 2 + 2 = 22 bytes
+struct header
+{
+    // "DOTNET_IPC_V1" with trailing 0s
+    char[16]   magic = { 0x68, 0x79, 0x84, 0x78, 0x69, 0x84, 0x95, 0x73, 0x80, 0x67, 0x95, 0x86, 0x49, 0x00, 0x00, 0x00 };
```

Comment:
This magic version "number" seems excessively long to me.  I would love some suggestions on a shorter form of this, that keeps things aligned so we don't have wasted padding bytes.

Reply:
```cpp
char magic[16] = "DOTNET_IPC_V1";
```

Reply:
As in I would rather use a magic version string that isn't 16 bytes long.  16 bytes seems excessive for what we are conveying with this byte array.